### ISSUE TITLE
This operation is only valid on generic types 

### ISSUE THREAD

**Issue by vascofernandes** [AutoMapper/AutoMapper#1389]
Hi,

I have the following code which worked with 4.2, but not with 5.0:

```csharp
    public class AutoMapperXYZ : Profile
    {
        public AutoMapperXYZ()
        {
            this.CreateMap<DateTimeOffset?, DateTime?>().ConvertUsing<NullableDateTimeOffsetConverter>();
        }
    }


    public class NullableDateTimeOffsetConverter : ITypeConverter<DateTimeOffset?, DateTime?>
    {
        public DateTime? Convert(DateTimeOffset? source, ResolutionContext context)
        {
            if (source.HasValue)
            {
                return source.Value.DateTime;
            }
            return default(DateTime?);
        }
    }
```

Throws the exception "This operation is only valid on generic types." in  AutoMapper\TypeExtensions.cs

Line 99:         public static Type[] GetGenericParameters(this Type type)
Line 100:        {
Line 101:                return type.GetGenericTypeDefinition().GetTypeInfo().GenericTypeParameters;
Line 102:        }
Line 103:

`type.GetGenericTypeDefinition()` throws the exception, with `type` being `NullableDateTimeOffsetConverter`.

**Comment by mobius7777** [comment]
I have this exact issue (and others too...). Mine with with a custom converter for DateTime? to DateTime?, which I know is weird, but it is doing something conditionally based on the source and destination models to adjust for the timezone. Regardless, my converter is called TimeZoneConverter and there is a nullable one too, called NullableTimeZoneConverter, with the interface for ITypeConverter, just like your NullableDateTimeOffsetConverter I would imagine. 

Now, the non-nullable one works, but the nullable one throws "This operation is only valid on generic types."

CreateMap<System.DateTime, System.DateTime>().ConvertUsing<TimeZoneConverter>();
CreateMap<System.DateTime?, System.DateTime?>().ConvertUsing<NullableTimeZoneConverter>();

I tried doing it this way as well:
CreateMap<System.DateTime?, System.DateTime?>().ConvertUsing(new NullableTimeZoneConverter());

However, that gave a different, and even stranger error : "Expression of type 'System.Nullable`1[System.Int32]' cannot be used for parameter of type 'System.Nullable`1[System.DateTime]'" 

So then I tried it this way:
CreateMap<System.DateTime?, System.DateTime?>().ConvertUsing((src, ctx) => new NullableTimeZoneConverter().Convert(src, ctx));

And got the same "Expression of type 'System.Nullable`1[System.Int32]' cannot be used for parameter of type 'System.Nullable`1[System.DateTime]'" error

This one is really strange since there is no integer involved at all. It just nullable datetime to nullable datetime, yet, somehow it thinks it is converting from a nullable int?
